### PR TITLE
Teuchos: Automatically enabling Tecuhos_ENABLE_THREAD_SAFE if you have Kokkos THREADS or OPENMP for the host

### PR DESCRIPTION
**Pull request by csiefer2**
Issue: #11921
Stakeholder feedback: Offline discussions with @crtrott as well as in #11921
Test: Manually verified for THREADS backend

### CODE REVIEW

**Comment by bartlettroscoe** [comment]
> Uhhh. _That_ isn't good. Try a restest and hope for the best?

@csiefer2, restart is not going to help.  I can't explain it but the new test `TrilinosInstallTests_simpleBuildAgainstTrilinos_by_package_build_tree` added in PR #11863 that was just merged is failing showing:

```
CMake Error in CMakeLists.txt:
  Imported target "Tpetra::all_libs" includes non-existent path

    "/scratch/trilinos/workspace/Trilinos_PR_cuda-11.4.2-uvm-off/pull_request_test/packages/kokkos-kernels/common"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

That test passed when run in the PR #11863.

If this failure can be reproduced locally, we can fix the CMakeLists.txt files to list this bad path (and other bad paths like I had to do in https://github.com/trilinos/Trilinos/pull/11863/commits/29fab02654a91f093ecbcecc325f7e9243ecde11 in PR #11863).

**Comment by bartlettroscoe** [comment]
@csiefer2, there is a good chance that commit https://github.com/trilinos/Trilinos/pull/11946/commits/21e643d7566c5afb5b882641378e9237a1681b70 will fix this.

File: packages/teuchos/core/test/MemoryManagement/ArrayView_MT_UnitTests_Decl.hpp [hunk 1]
```diff
@@ -202,7 +202,7 @@ static void share_arrayview_to_threads(ArrayView<int> shared_arrayview,
       ++cycle;
     }
   }
-  catch (DanglingReferenceError) {
+  catch (DanglingReferenceError&) {
```
Comment (bartlettroscoe) [inline packages/teuchos/core/test/MemoryManagement/ArrayView_MT_UnitTests_Decl.hpp:205]: 👍

**Review by bartlettroscoe — APPROVED** [review]
Looks like that should work.  Feel free to merge (once testing passes).  But I suggested some simplification of some IF() statements that you might want to consider.

File: packages/kokkos-kernels/common/CMakeLists.txt [hunk 2]
```diff
@@ -3,7 +3,3 @@ LIST(APPEND KK_INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/common/src)
 LIST(APPEND KK_INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/common/src/impl)
 LIST(APPEND KK_INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/common/src/tpls)
 LIST(APPEND KK_INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/common/unit_test)
-
```
Comment (ndellingwood) [inline packages/kokkos-kernels/common/CMakeLists.txt:6]: These kokkos-kernels changes should have had a matching PR put in to the kernels repo, @bartlettroscoe can you submit a PR with these changes?

Comment (ndellingwood) [reply packages/kokkos-kernels/common/CMakeLists.txt:6]: Matching patch to kokkos-kernels is up https://github.com/kokkos/kokkos-kernels/pull/1854

**Pull request merged**